File: frontend/src/types/workout.ts
/** A set that was actually performed. Sets the client skipped are simply absent. */
export interface LoggedSet {
  set_number: number;
  weight_kg: number | null;
  reps: number | null;
}

export interface WorkoutExercise {
  /** Id of the exercise within the plan — what a recorded set points back to. */
  id: string;
  name_es: string;
  image_path: string | null;
  sets: number;
  reps: string;
  rest_seconds: number | null;
  tempo: string | null;
  notes: string | null;
  superset_group: number | null;
  last_performed_on: string | null;
  last_sets: LoggedSet[];
}

export interface WorkoutDay {
  id: string;
  week_number: number;
  day_of_week_es: string;
  exercise_count: number;
  last_performed_on: string | null;
}

export interface WorkoutDayDetail {
  id: string;
  week_number: number;
  day_of_week_es: string;
  plan_title: string;
  exercises: WorkoutExercise[];
}

export interface LoggedSetInput {
  training_day_exercise_id: string | null;
  exercise_name: string;
  order_index: number;
  set_number: number;
  weight_kg: number | null;
  reps: number | null;
}

/**
 * A finished session, exactly as it travels to the server. It is also what gets
 * parked in the phone when there is no coverage, so it has to be self-contained.
 */
export interface WorkoutSessionInput {
  device_session_id: string;
  training_day_id: string;
  performed_on: string;
  notes: string | null;
  sets: LoggedSetInput[];
}

export interface WorkoutSet extends LoggedSet {
  exercise_name: string;
  order_index: number;
  target_reps: string | null;
}

export interface WorkoutSession {
  id: string;
  performed_on: string;
  day_of_week_es: string;
  plan_title: string | null;
  notes: string | null;
  sets: WorkoutSet[];
}

export interface WorkoutSessionSummary {
  id: string;
  performed_on: string;
  day_of_week_es: string;
  plan_title: string | null;
  exercise_count: number;
  set_count: number;
  total_volume_kg: number;
}

export interface ExercisePoint {
  performed_on: string;
  top_weight_kg: number | null;
  top_reps: number | null;
  total_volume_kg: number;
}

export interface ExerciseHistory {
  exercise_name: string;
  points: ExercisePoint[];
}

/**
 * A day worth painting on the calendar: one the client trained, one the routine
 * asked for, or both. Quiet days are simply not in the list.
 */
export interface TrainingCalendarDay {
  date: string;
  trained: boolean;
  planned: boolean;
  session_id: string | null;
  set_count: number | null;
  exercise_count: number | null;
}

export interface TrainingCalendar {
  days: TrainingCalendarDay[];
  /** Days with at least one session, not sessions: two in a day is one day. */
  trained_count: number;
  planned_count: number;
  missed_count: number;
  /** Null when the active plan has no start date, so it cannot be placed. */
  plan_starts_on: string | null;
  has_active_plan: boolean;
}

/** An exercise the client has logged at least once, offered to be charted. */
export interface TrainedExercise {
  exercise_id: string;
  exercise_name: string;
  session_count: number;
  last_performed_on: string;
  best_weight_kg: number | null;
}

File: backend/app/schemas/workout.py
import uuid
from datetime import date

from pydantic import BaseModel, Field

# A session sent from the gym may take hours to reach us, and a phone clock can
# be a day off in either direction. Anything older is someone filling in a diary.
MAX_BACKDATED_DAYS = 60
MAX_SETS_PER_SESSION = 300


class LoggedSetOut(BaseModel):
    """A set that was performed. Skipped sets are simply not there."""

    set_number: int
    weight_kg: float | None
    reps: int | None


class WorkoutExerciseOut(BaseModel):
    """An exercise as the training screen needs it: target and last time's numbers."""

    id: uuid.UUID
    name_es: str
    image_path: str | None
    sets: int
    reps: str
    rest_seconds: int | None
    tempo: str | None
    notes: str | None
    superset_group: int | None
    last_performed_on: date | None
    last_sets: list[LoggedSetOut]


class WorkoutDayOut(BaseModel):
    """A day of the routine, as offered to be trained."""

    id: uuid.UUID
    week_number: int
    day_of_week_es: str
    exercise_count: int
    last_performed_on: date | None


class WorkoutDayDetailOut(BaseModel):
    id: uuid.UUID
    week_number: int
    day_of_week_es: str
    plan_title: str
    exercises: list[WorkoutExerciseOut]


class LoggedSetIn(BaseModel):
    """One performed set, as the phone recorded it.

    `exercise_name` travels with it: it is what the client had on screen, and the
    trainer may have rewritten the routine while the phone had no coverage.
    """

    training_day_exercise_id: uuid.UUID | None = None
    exercise_name: str = Field(min_length=1, max_length=255)
    order_index: int = Field(ge=0)
    set_number: int = Field(ge=1, le=99)
    weight_kg: float | None = Field(default=None, ge=0, le=9999)
    reps: int | None = Field(default=None, ge=0, le=999)


class WorkoutSessionCreate(BaseModel):
    """A whole session, sent in one go once it is finished.

    `device_session_id` is generated on the phone so that resending — which is
    the normal outcome of training without coverage — cannot create a duplicate.
    """

    device_session_id: str = Field(min_length=8, max_length=64)
    training_day_id: uuid.UUID
    performed_on: date
    notes: str | None = Field(default=None, max_length=2000)
    sets: list[LoggedSetIn] = Field(min_length=1, max_length=MAX_SETS_PER_SESSION)


class WorkoutSetOut(LoggedSetOut):
    exercise_name: str
    order_index: int
    target_reps: str | None


class WorkoutSessionOut(BaseModel):
    """A session as it is read back, by the client and by the trainer alike."""

    id: uuid.UUID
    performed_on: date
    day_of_week_es: str
    plan_title: str | None
    notes: str | None
    sets: list[WorkoutSetOut]


class WorkoutSessionSummaryOut(BaseModel):
    """One line of history: enough to list sessions without loading every set."""

    id: uuid.UUID
    performed_on: date
    day_of_week_es: str
    plan_title: str | None
    exercise_count: int
    set_count: int
    total_volume_kg: float


class ExercisePointOut(BaseModel):
    """The heaviest set of a day, which is what a progression chart is made of."""

    performed_on: date
    top_weight_kg: float | None
    top_reps: int | None
    total_volume_kg: float


class ExerciseHistoryOut(BaseModel):
    exercise_name: str
    points: list[ExercisePointOut]


class TrainedExerciseOut(BaseModel):
    """One line per exercise the client has logged, to pick a progression from."""

    exercise_id: str
    exercise_name: str
    session_count: int
    last_performed_on: date
    best_weight_kg: float | None
